### PR TITLE
Refactored CRUD functions for AIRCRAFT, EMPLOYEE, and WORKSON

### PR DESCRIPTION
**Pull request by galenhwang**

### CODE REVIEW

File: airplanedb.py [hunk 1]
```diff
@@ -813,6 +1066,301 @@ def delete_gate(self, ap_id, g_id):
...
+            get_aircraft_query = """SELECT * FROM AIRCRAFT WHERE AC_ID = %d""" % (int(ac_id))
+        cursor = db.cursor()
+        try:
+            dataList = []
+            cursor.execute(get_aircraft_query)
+            if ac_id is None:
+                aircrafts = cursor.fetchall()
+            else:
+                aircrafts = cursor.fetchone()
+
```
Comment (jefftwin) [inline airplanedb.py]: After changing from jsonify to json.dump, this implementation for 

> cursor.fetchone()

doesnt work anymore. please look at the updated getAirport implementation

File: airplanedb.py [hunk 2]
```diff
@@ -813,6 +1066,301 @@ def delete_gate(self, ap_id, g_id):
...
+                'Number of Economy': aircraft[6],
+                'Number of Business': aircraft[7],
+                'Number of First Class': aircraft[8],
+                'Airport ID': aircraft[9]
+            }
+            data = json.dumps(ac_object, sort_keys=True, indent=4, separators=(',', ': '))
+        except Exception as err:
+            data = 'Update Aircraft Failed with error: {0}'.format(err)
+            db.rollback()
+
```
Comment (jefftwin) [inline airplanedb.py]: print out error with
`print(data)`
here too

File: airplanedb.py [hunk 3]
```diff
@@ -813,6 +1066,301 @@ def delete_gate(self, ap_id, g_id):
...
+#   function: delete_aircraft
+#   description: delete an aircraft from table AIRCRAFT
+#==============================================================================
+    def delete_aircraft(self, ac_id):
+        db = MySQLdb.connect(host=self.host,
+                            user=self.user,
+                            passwd=self.pw,
+                            db=self.db)
+
+        delete_aircraft_query = """DELETE FROM AIRCRAFT WHERE AC_ID = %d""" % (int (ac_id))
```
Comment (jefftwin) [inline airplanedb.py]: not sure if the space in
`int (ac_id)`
would cause an error. delete the space

File: airplanedb.py [hunk 4]
```diff
@@ -813,6 +1066,301 @@ def delete_gate(self, ap_id, g_id):
...
+                            passwd=self.pw,
+                            db=self.db)
+
+        add_employee_query = """INSERT INTO EMPLOYEE (E_HOURS, E_TYPE, E_WAGE)
+                                VALUES (%.2f, '%s', %.2f)""" % (float(hours),
+                                emp_type, float(wage))
+
+        cursor = db.cursor()
+        employee = {
+            'ID': str(cursor.lastrowid),
```
Comment (jefftwin) [inline airplanedb.py]: is employeeID supposed to be string? should be int so leave it as an int in the json object

File: airplanedb.py [hunk 5]
```diff
@@ -850,6 +1398,65 @@ def get_schedule_for_itinerary(self, i_id):
...
+            data = 'Add Workson Failed with error: {0}'.format(err)
+            db.rollback()
+
+        cursor.close()
+        db.close()
+        return data
+
+#==============================================================================
+#   function: delete_workson
+#   description: delete a workson relation instance from table WORKSON
```
Comment (jefftwin) [inline airplanedb.py]: return comment

**Review by jefftwin — CHANGES_REQUESTED** [review]
remember to add return comment and also have print statements when fail so we can check the log for exact failures. 

MAJOR ERRORS: check the get Aircraft function. the cursor.fetchone() doesn't work with json.dumps while cursor.fetchall(). This leads me to think you didn't full test each path. Please test each path and make sure everything works O=

File: airplane.py [hunk 6]
```diff
@@ -216,6 +216,116 @@ def update_airport():
     return data
 
 
+# Delete a gate route
+@app.route('/gate/delete')
```
Comment (maknoon) [inline airplane.py]: You can remove this function; it is already implemented [here](https://github.com/maknoon/airdb/blob/dev_connie/airplane.py#L260)

File: airplane.py [hunk 7]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+def delete_gate():
+    apid = request.args.get('ap_id')
+    gid = request.args.get('g_id')
+    data = airdb.delete_gate(apid, gid)
+
+    return data
+
+# Add new aircraft to table AIRCRAFT
+@app.route('/aircraft/new')
```
Comment (maknoon) [inline airplane.py]: Refactor `/aircraft/new`, `/aircraft/get`, `/aircraft/update` & `/aircraft/delete` into one endpoint `/aircraft`, see [this example](https://github.com/maknoon/airdb/blob/dev_connie/airplane.py#L211) for how to do it

File: airplane.py [hunk 8]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+
+# Delete aircraft in table AIRCRAFT
+@app.route('/aircraft/delete')
+def delete_aircraft():
+	data = airdb.delete_aircraft(request.args.get('id'))
+
+	return data
+
+# Add new employee to table EMPLOYEE
+@app.route('/employee/new')
```
Comment (maknoon) [inline airplane.py]: please do the same RESTful refactoring here. see comment for refactoring `/aircraft`

File: airplane.py [hunk 9]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+
+# Delete employee in table EMPLOYEE
+@app.route('/employee/delete')
+def delete_employee():
+	data = airdb.delete_employee(request.args.get('id'))
+
+	return data
+
+# Get schedule of ITINERARY
+@app.route('/schedule/getForItinerary')
```
Comment (maknoon) [inline airplane.py]: This function can be removed; it is already included [here](https://github.com/maknoon/airdb/blob/dev_connie/airplane.py#L274)

File: airplane.py [hunk 10]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+def get_schedule_for_itinerary():
+    iid = request.args.get('id')
+    data = airdb.get_schedule_for_itinerary(iid)
+
+    return data
+
+# Add a new workson relation to table WORKSON
+@app.route('/workson/new')
```
Comment (maknoon) [inline airplane.py]: please do the same RESTful refactoring here. see comment for refactoring `/aircraft`

File: airplane.py [hunk 11]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+
+# Delete a workson relation in table Workson
+@app.route('/workson/delete')
+def delete_workson():
+    data = airdb.delete_workson(request.args.get('e_id'), request.args.get('f_id'))
+
+    return data
+
+# Add a schedule route
+@app.route('/schedule/add')
```
Comment (maknoon) [inline airplane.py]: Can remove this function; it is already being implemented [here](https://github.com/maknoon/airdb/blob/dev_connie/airplane.py#L269)

File: airplane.py [hunk 12]
```diff
@@ -216,6 +216,116 @@ def update_airport():
...
+def add_schedule():
+    iid = request.args.get('i_id')
+    fid = request.args.get('f_id')
+    data = airdb.add_schedule(iid, fid)
+
+    return data
+
+# Delete a schedule route
+@app.route('/schedule/delete')
```
Comment (maknoon) [inline airplane.py]: Can remove this function; it is already being implemented [here](https://github.com/maknoon/airdb/blob/dev_connie/airplane.py#L269)

File: airplanedb.py [hunk 13]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                                VALUES ('%s', '%s', %.2f, '%s', '%s', %d, %d, %d, '%s')""" % (
+                                status, make, float(mileage), datecreated, lastmaintained,
+                                int(economy), int(business), int(firstclass), airport)
+
+        cursor = db.cursor()
+        try:
+            cursor.execute(add_aircraft_query)
+            db.commit()
+            aircraft = {
+                'ID': cursor.lastrowid,
```
Comment (maknoon) [inline airplanedb.py]: all JSON object props should be in `snake_case`
e.g. `Airport ID` -> `airport_id`
`Date Created` -> `date_created`
`Number of Economy` -> `num_economy`

File: airplanedb.py [hunk 14]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+            get_aircraft_query = """SELECT * FROM AIRCRAFT WHERE AC_ID = %d""" % (int(ac_id))
+        cursor = db.cursor()
+        try:
+            dataList = []
+            cursor.execute(get_aircraft_query)
+            if ac_id is None:
+                aircrafts = cursor.fetchall()
+                for aircraft in aircrafts:
+                    ac_object = {
+                        'ID': aircraft[0],
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 15]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                        'Number of Economy': aircraft[6],
+                        'Number of Business': aircraft[7],
+                        'Number of First Class': aircraft[8],
+                        'Airport ID': aircraft[9]
+                    }
+                    dataList.append(ac_object)
+            else:
+                aircrafts = cursor.fetchone()
+                ac_object = {
+                    'ID': aircrafts[0],
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 16]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                                                                            int(ac_id))
+        cursor = db.cursor()
+        try:
+            cursor.execute(update_aircraft_query)
+            db.commit()
+            get_aircraft_query = """SELECT * FROM AIRCRAFT WHERE AC_ID = %d""" % (int(ac_id))
+            cursor.execute(get_aircraft_query)
+            aircraft = cursor.fetchone()
+            ac_object = {
+                'ID': aircraft[0],
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 17]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                            passwd=self.pw,
+                            db=self.db)
+
+        add_employee_query = """INSERT INTO EMPLOYEE (E_HOURS, E_TYPE, E_WAGE)
+                                VALUES (%.2f, '%s', %.2f)""" % (float(hours),
+                                emp_type, float(wage))
+
+        cursor = db.cursor()
+        employee = {
+            'ID': cursor.lastrowid,
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 18]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                                    FROM EMPLOYEE E, WORKSON W WHERE
+                                    E.E_ID = W.E_ID AND W.F_ID = %d""" % (int(f_id))
+        cursor = db.cursor()
+        try:
+            dataList = []
+            cursor.execute(get_employee_query)
+            employees = cursor.fetchall()
+            for e in employees:
+                employee = {
+                    'ID': e
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 19]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+                                    FROM FLIGHT F, WORKSON W WHERE
+                                    F.F_ID = W.F_ID AND W.E_ID = %d""" % (int(e_id))
+        cursor = db.cursor()
+        try:
+            dataList = []
+            cursor.execute(get_flight_query)
+            flights = cursor.fetchall()
+            for f in flights:
+                flight = {
+                    'ID': f
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 20]
```diff
@@ -1066,6 +1066,327 @@ def delete_gate(self, ap_id, g_id):
...
+    def delete_employee(self, e_id):
+        db = MySQLdb.connect(host=self.host,
+                            user=self.user,
+                            passwd=self.pw,
+                            db=self.db)
+
+        delete_employee_query = """DELETE FROM EMPLOYEE WHERE E_ID = %d""" % (int (e_id))
+        cursor = db.cursor()
+        deleted_employee_id = {
+            'ID': e_id
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

File: airplanedb.py [hunk 21]
```diff
@@ -1103,6 +1424,69 @@ def get_schedule_for_itinerary(self, i_id):
...
+                            user=self.user,
+                            passwd=self.pw,
+                            db=self.db)
+
+        add_workson_query = """INSERT INTO WORKSON VALUES (%d, %d)""" % (int(e_id),
+                                int(f_id))
+
+        cursor = db.cursor()
+        workson = {
+            'E_ID': e_id,
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`
`E_ID` & `F_ID` should be `employee_id` & `flight_id`, respectively

File: airplanedb.py [hunk 22]
```diff
@@ -1103,6 +1424,69 @@ def get_schedule_for_itinerary(self, i_id):
...
+        db = MySQLdb.connect(host=self.host,
+                            user=self.user,
+                            passwd=self.pw,
+                            db=self.db)
+
+        delete_workson_query = """DELETE FROM WORKSON WHERE E_ID = %d and
+                                F_ID = %d""" % (int (e_id), int (f_id))
+        cursor = db.cursor()
+        deleted_workson = {
+            'E_ID': e_id,
```
Comment (maknoon) [inline airplanedb.py]: use `snake_case`

**Review by maknoon — CHANGES_REQUESTED** [review]
Looking good so far, please use `snake_case` for naming JSON object properties!

**Review by katieli16 — CHANGES_REQUESTED** [review]
Can you write a get_works_on function that returns the entire works_on table?

**Comment by maknoon** [comment]
~~@galenhwang can you delete `.DS_Store` from your PR (`git rm .DS_Store`) and add it to the `.gitignore` please~~ nvm I did it, just pull from this branch

**Review by maknoon — APPROVED** [review]

**Review by jefftwin — APPROVED** [review]

**Pull request merged**